File: app/src/main/java/com/simon/harmonichackernews/data/StackExchangeInfo.java
package com.simon.harmonichackernews.data;

public class StackExchangeInfo {

    public String title;
    public String author;
    public String questionText;
    public String[] tags;
    public String site;
    public int score;
    public int answerCount;
    public int viewCount;
    public boolean isAnswered;
    public boolean hasAcceptedAnswer;

    public String formatScore() {
        return LinkPreviewFormatUtils.formatCount(score, "point", "points");
    }

    public String formatAnswerCount() {
        return LinkPreviewFormatUtils.formatCount(answerCount, "answer", "answers");
    }

    public String formatViewCount() {
        return LinkPreviewFormatUtils.formatCount(viewCount, "view", "views");
    }

    public String formatAnswerState() {
        if (hasAcceptedAnswer) {
            return "Accepted answer";
        }

        if (isAnswered) {
            return "Answered";
        }

        return "Unanswered";
    }

    public String formatTags() {
        if (tags == null || tags.length == 0) {
            return null;
        }

        StringBuilder builder = new StringBuilder();
        for (int i = 0; i < tags.length; i++) {
            if (i > 0) {
                builder.append(", ");
            }
            builder.append(tags[i]);
        }
        return builder.toString();
    }

    public String formatBy() {
        if (questionText != null) {
            return questionText;
        }

        if (author == null) {
            return site;
        }

        return author + " on " + site;
    }

    public String formatAuthor() {
        if (author == null) {
            return site;
        }

        return author;
    }
}

File: app/src/main/java/com/simon/harmonichackernews/data/StackExchangeInfo.kt
package com.simon.harmonichackernews.data

import com.simon.harmonichackernews.data.LinkPreviewFormatUtils.formatCount

class StackExchangeInfo {
    var title: String? = null
    var author: String? = null
    var questionText: String? = null
    var tags: Array<String?>? = null
    var site: String? = null
    var score: Int = 0
    var answerCount: Int = 0
    var viewCount: Int = 0
    var isAnswered: Boolean = false
    var hasAcceptedAnswer: Boolean = false

    fun formatScore(): String {
        return formatCount(score, "point", "points")
    }

    fun formatAnswerCount(): String {
        return formatCount(answerCount, "answer", "answers")
    }

    fun formatViewCount(): String {
        return formatCount(viewCount, "view", "views")
    }

    fun formatAnswerState(): String {
        if (hasAcceptedAnswer) {
            return "Accepted answer"
        }

        if (isAnswered) {
            return "Answered"
        }

        return "Unanswered"
    }

    fun formatTags(): String? {
        val currentTags = tags
        if (currentTags.isNullOrEmpty()) {
            return null
        }

        val builder = StringBuilder()
        for (i in currentTags.indices) {
            if (i > 0) {
                builder.append(", ")
            }
            builder.append(currentTags[i])
        }
        return builder.toString()
    }

    fun formatBy(): String? {
        if (questionText != null) {
            return questionText
        }

        if (author == null) {
            return site
        }

        return author + " on " + site
    }

    fun formatAuthor(): String? {
        if (author == null) {
            return site
        }

        return author
    }
}
